web/script.js

```diff
@@ -0,0 +1,355 @@
+            // limit scrolling by scrolling to a point where the area of interest is at the top
+            newViewportY = newViewportYmax;
+            newViewportY -= offset;
+            if (newViewportY < newViewportYmin) {
+                // constrained on both directions
+                newViewportY = (newViewportYmin + newViewportYmax) / 2;
+            }
+        } else { // currently scrolled above (or at least, not below) the area we want to scroll to
+            // scroll to a point where the area of interest is at the bottom
+            newViewportY = newViewportYmin;
+            newViewportY += offset;
+            if (newViewportY > newViewportYmax) {
+                newViewportY = (newViewportYmin + newViewportYmax) / 2;
+            }
+        }
+    } else {
+        let newViewportYmax = elem1y;
+        let newViewportYmin = elem1y + elem1height - viewportHeight;
+        // "ideal" viewport dimensions for showing all elements in one view
+        let idealViewportY = elemMinY;
+        let idealViewportHeight = elemMaxY + elemMaxYheight - elemMinY;
+        // translate the viewport around elem1y according to location of other elements
+        newViewportY = elem1y + (idealViewportY - elem1y) / idealViewportHeight * viewportHeight;
+        if (newViewportY >= newViewportYmax - offset) {
+            newViewportY = newViewportYmax - offset;
+        }
+        if (newViewportY <= newViewportYmin + offset) {
+            newViewportY = newViewportYmin + offset;
+        }
+    }
+
+    window.scrollTo({
+        top: newViewportY,
+        behavior: 'smooth',
+    });
+}
+
+function highlightBranchClearAll() {
+    for (let elem of document.getElementsByClassName('branch-indicator-wrapper')) {
+        elem.classList.remove('branch-highlight-stay');
+        elem.classList.remove('branch-highlight-temp');
+    }
+}
+
+function highlightBranchStay(className) {
+    for (let elem of document.getElementsByClassName(className)) {
+        elem.parentNode.classList.add('branch-highlight-stay');
+    }
+}
+
+function highlightBranchTemp(className) {
+    for (let elem of document.getElementsByClassName(className)) {
+        elem.parentNode.classList.add('branch-highlight-temp');
+    }
+}
+
+function highlightBranchTempClear(className) {
+    for (let elem of document.getElementsByClassName(className)) {
+        elem.parentNode.classList.remove('branch-highlight-temp');
+    }
+}
+
+function onClickBranchOrigin(elem) {
+    let branchTargetElem = document.getElementById(elem.dataset.branchTarget);
+    if (branchTargetElem) {
+        scrollIntoView(branchTargetElem, [elem]);
+    }
+    highlightBranchClearAll();
+    highlightBranchStay(elem.dataset.branchesClass);
+}
+
+function onClickBranchTarget(elem) {
+    // get all branch origins that branch to this branch target
+    let branchOriginsElems = [];
+    for (let branchIndicatorElem of document.getElementsByClassName(elem.dataset.branchesClass)) {
```

Comment:
I think you can simplify the HTML a bit by using `document.querySelectorAll('data-rotation="' + elem.dataset.rotation + '"')` ? (And similar, to avoid having to add many new data attributes/classes?)

Reply:
I'm avoiding relying on data-rotation since this "rotation" stuff is purely intended for colors rather than describing data. What I mean is I prefer verbosity over over-engineered reused data that produces hard-to-fix bugs

Also it's not like we are restrained on how much data we can send through localhost (a html diff with default max func size is ~200ko, even for decomp.me it would be manageable)